**Q1. Problem Statement: Linear Programming using PuLP**

Suppose you are a dietician and are responsible to advise one of your customers on the best possible food plan he should follow in order to attain the optimum nutrition.  However, there are some restrictions in terms of the budget and the variety of food that needs to be included in the diet plan. The cost should be minimal, at the same time, the nutritional value derived from the combination of different food items should be maximum, considering the maximum and minimum constraints given in the data.

Load the 'deit.csv' dataset into a DataFrame and perform the following tasks:
1.	Download and import PuLP and other required libraries
2.	Formulate the optimization problem using LpProblem() method
3.	Get a list of all the food items present in the DataFrame
4.	Create a separate dictionary for each of these columns - 'Price/Serving', 'Calories', 'Cholesterol (mg)', 'Total_Fat (g)', 'Sodium (mg)', 'Carbohydrates (g)', 'Dietary_Fiber (g)', 'Protein (g)', 'Vit_A (IU)', 'Vit_C (IU)', 'Calcium (mg)', and 'Iron (mg)' using zip() function
5.	Create a dictionary called 'food_vars' to store the referenced variables for food items with the 'food' list
6.	Add an objective function (lpSum() ) to the LPP
7.	Add the maximum and minimum constraints for each of the nutritional values present in the DataFrame
8.	Run the solver algorithm to solve the problem
9.	Print the variables after the solution
10.	Based on the results, prepare an optimal diet plan


**Step-1:** Downloading and importing pulp and other required libraries.

In [ ]:
!pip install pulp

In [ ]:
import pulp
from pulp import *
import pandas as pd
import warnings
warnings.filterwarnings('always')
warnings.filterwarnings('ignore')

**Step-2:** Formulating the optimization problem.


In [ ]:
prob = LpProblem('Diet_Problem', LpMinimize)

**Step-3:** Loading the dataset into a DataFrame.

In [ ]:
df=pd.read_csv('/content/sample_data/diet.csv')
df.head()

Foods Serving Size  Calories  Cholesterol (mg)  \
0      Frozen Broccoli    10 Oz Pkg      73.8               0.0   
1          Frozen Corn      1/2 Cup      72.2               0.0   
2  Raw Lettuce Iceberg       1 Leaf       2.6               0.0   
3       Baked Potatoes      1/2 Cup     171.5               0.0   
4                 Tofu    1/4 block      88.2               0.0   

   Total_Fat (g)  Sodium (mg)  Carbohydrates (g)  Dietary_Fiber (g)  \
0            0.8         68.2               13.6                8.5   
1            0.6          2.5               17.1                2.0   
2            0.0          1.8                0.4                0.3   
3            0.2         15.2               39.9                3.2   
4            5.5          8.1                2.2                1.4   

   Protein (g)  Vit_A (IU)  Vit_C (IU)  Calcium (mg)  Iron (mg)  \
0          8.0      5867.4       160.2         159.0        2.3   
1          2.5       106.6         5.2           3.3        0.3   
2          0.2        66.0         0.8           3.8        0.1   
3          3.7         0.0        15.6          22.7        4.3   
4          9.4        98.6         0.1         121.8        6.2   

   Price/Serving ($)  
0               0.48  
1               0.54  
2               0.06  
3               0.18  
4               0.93

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Foods              17 non-null     object 
 1   Serving Size       17 non-null     object 
 2   Calories           17 non-null     float64
 3   Cholesterol (mg)   17 non-null     float64
 4   Total_Fat (g)      17 non-null     float64
 5   Sodium (mg)        17 non-null     float64
 6   Carbohydrates (g)  17 non-null     float64
 7   Dietary_Fiber (g)  17 non-null     float64
 8   Protein (g)        17 non-null     float64
 9   Vit_A (IU)         17 non-null     float64
 10  Vit_C (IU)         17 non-null     float64
 11  Calcium (mg)       17 non-null     float64
 12  Iron (mg)          17 non-null     float64
 13  Price/Serving ($)  17 non-null     float64
dtypes: float64(12), object(2)
memory usage: 2.0+ KB


**Step-4:** Gettting the list of all the food items present in the DataFrame.

In [ ]:
#List of food items
food = list(df.Foods)

#The list of items
count=pd.Series(range(1,len(food)+1))
print('List of different food items is here follows: -')
food_s = pd.Series(food)

#Convert to data frame
f_frame = pd.concat([count,food_s],axis=1,keys=['S.No','Food Items'])
f_frame

List of different food items is here follows: -


S.No              Food Items
0      1         Frozen Broccoli
1      2             Frozen Corn
2      3     Raw Lettuce Iceberg
3      4          Baked Potatoes
4      5                    Tofu
5      6         Roasted Chicken
6      7      Spaghetti W/ Sauce
7      8               Raw Apple
8      9                  Banana
9     10             Wheat Bread
10    11             White Bread
11    12         Oatmeal Cookies
12    13               Apple Pie
13    14          Scrambled Eggs
14    15          Turkey Bologna
15    16        Beef Frankfurter
16    17  Chocolate Chip Cookies

**Step-5:** Create a separate dictionary for each of these columns - 'Price/Serving', 'Calories', 'Cholesterol (mg)', 'Total_Fat (g)', 'Sodium (mg)', 'Carbohydrates (g)', 'Dietary_Fiber (g)', 'Protein (g)', 'Vit_A (IU)', 'Vit_C (IU)', 'Calcium (mg)', and 'Iron (mg)' using zip() function.

In [ ]:
# Create a dictinary of costs for all food items
costs = dict(zip(food,df['Price/Serving ($)']))

#Create a dictionary of calories for all items of food
calories = dict(zip(food,df['Calories']))

#Create a dictionary of cholesterol for all items of food
chol = dict(zip(food,df['Cholesterol (mg)']))

#Create a dictionary of total fat for all items of food
fat = dict(zip(food,df['Total_Fat (g)']))

#Create a dictionary of sodium for all items of food
sodium = dict(zip(food,df['Sodium (mg)']))

#Create a dictionary of carbohydrates for all items of food
carbs = dict(zip(food,df['Carbohydrates (g)']))

#Create a dictionary of dietary fiber for all items of food
fiber = dict(zip(food,df['Dietary_Fiber (g)']))

#Create a dictionary of protein for all food items
protein = dict(zip(food,df['Protein (g)']))

#Create a dictionary of vitamin A for all food items
vit_A = dict(zip(food,df['Vit_A (IU)']))

#Create a dictionary of vitamin C for all food items
vit_C = dict(zip(food,df['Vit_C (IU)']))

#Create a dictionary of calcium for all food items
calcium = dict(zip(food,df['Calcium (mg)']))

#Create a dictionary of iron for all food items
iron = dict(zip(food,df['Iron (mg)']))

**Step-6:** Running one of the dictionaries to see how these look like.

In [ ]:
#We just run one of the dictionaries to see how these look like
iron

{'Frozen Broccoli': 2.3,
 'Frozen Corn': 0.3,
 'Raw Lettuce Iceberg': 0.1,
 ' Baked Potatoes': 4.3,
 'Tofu': 6.2,
 'Roasted Chicken': 1.8,
 'Spaghetti W/ Sauce': 2.3,
 'Raw Apple': 0.2,
 'Banana': 0.4,
 'Wheat Bread': 0.7,
 'White Bread': 0.8,
 'Oatmeal Cookies': 0.5,
 'Apple Pie': 0.1,
 'Scrambled Eggs': 0.7,
 'Turkey Bologna': 0.4,
 'Beef Frankfurter': 0.6,
 'Chocolate Chip Cookies': 0.4}

**Step-7:**  Creating a dictionary called 'food_vars' to store the referenced variables for food items with the 'food' list.


In [ ]:
# A dictionary called 'food_vars' is created to contain the referenced Variables
food_vars = LpVariable.dicts("Food",food,lowBound=0,cat='Continuous')

In [ ]:
food_vars

{'Frozen Broccoli': Food_Frozen_Broccoli,
 'Frozen Corn': Food_Frozen_Corn,
 'Raw Lettuce Iceberg': Food_Raw_Lettuce_Iceberg,
 ' Baked Potatoes': Food__Baked_Potatoes,
 'Tofu': Food_Tofu,
 'Roasted Chicken': Food_Roasted_Chicken,
 'Spaghetti W/ Sauce': Food_Spaghetti_W__Sauce,
 'Raw Apple': Food_Raw_Apple,
 'Banana': Food_Banana,
 'Wheat Bread': Food_Wheat_Bread,
 'White Bread': Food_White_Bread,
 'Oatmeal Cookies': Food_Oatmeal_Cookies,
 'Apple Pie': Food_Apple_Pie,
 'Scrambled Eggs': Food_Scrambled_Eggs,
 'Turkey Bologna': Food_Turkey_Bologna,
 'Beef Frankfurter': Food_Beef_Frankfurter,
 'Chocolate Chip Cookies': Food_Chocolate_Chip_Cookies}

In [ ]:
prob += lpSum([costs[i]*food_vars[i] for i in food])
prob

Diet_Problem:
MINIMIZE
0.48*Food_Apple_Pie + 0.45*Food_Banana + 0.81*Food_Beef_Frankfurter + 0.09*Food_Chocolate_Chip_Cookies + 0.48*Food_Frozen_Broccoli + 0.54*Food_Frozen_Corn + 0.27*Food_Oatmeal_Cookies + 0.72*Food_Raw_Apple + 0.06*Food_Raw_Lettuce_Iceberg + 2.52*Food_Roasted_Chicken + 0.33*Food_Scrambled_Eggs + 2.34*Food_Spaghetti_W__Sauce + 0.93*Food_Tofu + 0.45*Food_Turkey_Bologna + 0.15*Food_Wheat_Bread + 0.18*Food_White_Bread + 0.18*Food__Baked_Potatoes + 0.0
VARIABLES
Food_Apple_Pie Continuous
Food_Banana Continuous
Food_Beef_Frankfurter Continuous
Food_Chocolate_Chip_Cookies Continuous
Food_Frozen_Broccoli Continuous
Food_Frozen_Corn Continuous
Food_Oatmeal_Cookies Continuous
Food_Raw_Apple Continuous
Food_Raw_Lettuce_Iceberg Continuous
Food_Roasted_Chicken Continuous
Food_Scrambled_Eggs Continuous
Food_Spaghetti_W__Sauce Continuous
Food_Tofu Continuous
Food_Turkey_Bologna Continuous
Food_Wheat_Bread Continuous
Food_White_Bread Continuous
Food__Baked_Potatoes Continuous

**Step-9:** Adding the maximum and minumum constraints for each of the nutritional values.

In [ ]:
lpSum([food_vars[i]*calories[i] for i in food])

67.2*Food_Apple_Pie + 104.9*Food_Banana + 141.8*Food_Beef_Frankfurter + 78.1*Food_Chocolate_Chip_Cookies + 73.8*Food_Frozen_Broccoli + 72.2*Food_Frozen_Corn + 81.0*Food_Oatmeal_Cookies + 81.4*Food_Raw_Apple + 2.6*Food_Raw_Lettuce_Iceberg + 277.4*Food_Roasted_Chicken + 99.6*Food_Scrambled_Eggs + 358.2*Food_Spaghetti_W__Sauce + 88.2*Food_Tofu + 56.4*Food_Turkey_Bologna + 65.0*Food_Wheat_Bread + 65.0*Food_White_Bread + 171.5*Food__Baked_Potatoes + 0.0

In [ ]:
df[['Foods','Calories']]

Foods  Calories
0          Frozen Broccoli      73.8
1              Frozen Corn      72.2
2      Raw Lettuce Iceberg       2.6
3           Baked Potatoes     171.5
4                     Tofu      88.2
5          Roasted Chicken     277.4
6       Spaghetti W/ Sauce     358.2
7                Raw Apple      81.4
8                   Banana     104.9
9              Wheat Bread      65.0
10             White Bread      65.0
11         Oatmeal Cookies      81.0
12               Apple Pie      67.2
13          Scrambled Eggs      99.6
14          Turkey Bologna      56.4
15        Beef Frankfurter     141.8
16  Chocolate Chip Cookies      78.1

In [ ]:
prob += lpSum([food_vars[x]*calories[x] for x in food]) >= 800, "CaloriesMinimum"
prob += lpSum([food_vars[x]*calories[x] for x in food]) <= 1300, "CaloriesMaximum"
prob

Diet_Problem:
MINIMIZE
0.48*Food_Apple_Pie + 0.45*Food_Banana + 0.81*Food_Beef_Frankfurter + 0.09*Food_Chocolate_Chip_Cookies + 0.48*Food_Frozen_Broccoli + 0.54*Food_Frozen_Corn + 0.27*Food_Oatmeal_Cookies + 0.72*Food_Raw_Apple + 0.06*Food_Raw_Lettuce_Iceberg + 2.52*Food_Roasted_Chicken + 0.33*Food_Scrambled_Eggs + 2.34*Food_Spaghetti_W__Sauce + 0.93*Food_Tofu + 0.45*Food_Turkey_Bologna + 0.15*Food_Wheat_Bread + 0.18*Food_White_Bread + 0.18*Food__Baked_Potatoes + 0.0
SUBJECT TO
CaloriesMinimum: 67.2 Food_Apple_Pie + 104.9 Food_Banana
 + 141.8 Food_Beef_Frankfurter + 78.1 Food_Chocolate_Chip_Cookies
 + 73.8 Food_Frozen_Broccoli + 72.2 Food_Frozen_Corn + 81 Food_Oatmeal_Cookies
 + 81.4 Food_Raw_Apple + 2.6 Food_Raw_Lettuce_Iceberg
 + 277.4 Food_Roasted_Chicken + 99.6 Food_Scrambled_Eggs
 + 358.2 Food_Spaghetti_W__Sauce + 88.2 Food_Tofu + 56.4 Food_Turkey_Bologna
 + 65 Food_Wheat_Bread + 65 Food_White_Bread + 171.5 Food__Baked_Potatoes
 >= 800

CaloriesMaximum: 67.2 Food_Apple_Pie + 104.9

In [ ]:
#Carbohydrates' constraint
prob += lpSum([food_vars[x]*carbs[x] for x in food]) >= 130, "CarbsMinimum"
prob += lpSum([food_vars[x]*carbs[x] for x in food]) <= 200, "CarbsMaximum"

#Fat's constraint
prob += lpSum([food_vars[x]*fat[x] for x in food]) >= 20, "FatsMinimum"
prob += lpSum([food_vars[x]*fat[x] for x in food]) <= 50, "FatsMaximum"

#Protein's constraint
prob += lpSum([food_vars[x]*protein[x] for x in food]) >= 100, "ProteinsMinimum"
prob += lpSum([food_vars[x]*protein[x] for x in food]) <= 150, "ProteinsMaximum"

#Vit_A constraint
prob += lpSum([food_vars[x]*vit_A[x] for x in food]) >= 1000, "Vit_A_Minimum"
prob += lpSum([food_vars[x]*vit_A[x] for x in food]) <= 10000, "Vit_A_Maximum"

**Step-10:** Running the Solver Algorithm.

In [ ]:
prob.solve()

1

In [ ]:
prob.solver

In [ ]:
LpStatus[prob.status]

'Optimal'

**Step-11:** Printing of the variables after solvingthe problem.

In [ ]:
# prob.variables()[1].varValue()

for var in prob.variables():
    print(f'Variable name: {var.name} , Variable value : {var.value()}\n')

print('\n')
print('*'*100)
print('\n')

#We can also see the slack variables of the constraints
for name, con in prob.constraints.items():
    print(f'constraint name:{name}, constraint value:{con.value()}\n')

print('*'*100)
print('\n')

## OBJECTIVE VALUE
print(f'OBJECTIVE VALUE IS: {round(prob.objective.value(),2)}')

Variable name: Food_Apple_Pie , Variable value : 0.0

Variable name: Food_Banana , Variable value : 0.0

Variable name: Food_Beef_Frankfurter , Variable value : 0.0

Variable name: Food_Chocolate_Chip_Cookies , Variable value : 0.0

Variable name: Food_Frozen_Broccoli , Variable value : 1.4079473

Variable name: Food_Frozen_Corn , Variable value : 0.0

Variable name: Food_Oatmeal_Cookies , Variable value : 0.0

Variable name: Food_Raw_Apple , Variable value : 0.0

Variable name: Food_Raw_Lettuce_Iceberg , Variable value : 0.0

Variable name: Food_Roasted_Chicken , Variable value : 1.2329408

Variable name: Food_Scrambled_Eggs , Variable value : 4.0165704

Variable name: Food_Spaghetti_W__Sauce , Variable value : 0.0

Variable name: Food_Tofu , Variable value : 0.0

Variable name: Food_Turkey_Bologna , Variable value : 0.0

Variable name: Food_Wheat_Bread , Variable value : 0.0

Variable name: Food_White_Bread , Variable value : 0.0

Variable name: Food__Baked_Potatoes , Variable value 

**Result:** The optimal value of the Diet is $5.58. Which includes the following food items:

--> 2.64 servings of baked potatoes

--> 4.02 servings of scrambled eggs

--> 1.23 servings of Roasted chicken

--> 1.41 servings of Frozen broccoli

